src/docgen/transpile/csharp.ts

```diff
@@ -0,0 +1,226 @@
+import * as Case from 'case';
+import * as reflect from 'jsii-reflect';
+import * as transpile from './transpile';
+
+export class CSharpTranspile extends transpile.TranspileBase {
+  constructor() {
+    super(transpile.Language.CSHARP);
+  }
+
+  public moduleLike(
+    moduleLike: reflect.ModuleLike,
+  ): transpile.TranspiledModuleLike {
+    const csharpPackage: string = moduleLike.targets?.dotnet?.namespace;
+
+    // if this is a submodule, we need to break the package name down into the
+    // parent name and the submodule. we also allow submodules not to have
+    // explicit target names, in which case we need to append the pascal-cased
+    // submodule name to the parent package name.
+    if (moduleLike instanceof reflect.Submodule) {
+      const parent = this.getParentModule(moduleLike);
+      const parentFqn = parent.targets?.dotnet?.namespace;
+
+      // if the submodule does not explicitly define a dotnet package name, we need to deduce it from the parent
+      const submoduleCSharpPackage = csharpPackage ?? `${parentFqn}.${Case.pascal(moduleLike.name)}`;
+      return { name: parentFqn, submodule: submoduleCSharpPackage };
+    }
+
+    return { name: csharpPackage };
+  }
+
+  public type(type: reflect.Type): transpile.TranspiledType {
+    const submodule = this.findSubmodule(type);
+    const moduleLike = this.moduleLike(submodule ? submodule : type.assembly);
+
+    const fqn = [];
+
+    let namespace = type.namespace;
+    if (namespace) {
+      fqn.push(moduleLike.name);
+      fqn.push(namespace);
+    } else {
+      fqn.push(moduleLike.name);
+    }
+    fqn.push(type.name);
+
+    return {
+      fqn: fqn.join('.'),
+      name: type.name,
+      namespace: namespace,
+      module: moduleLike.name,
+      submodule: moduleLike.submodule,
+      source: type,
+      language: this.language,
+    };
+  }
+
+  public callable(callable: reflect.Callable): transpile.TranspiledCallable {
+    const type = this.type(callable.parentType);
+    const isInitializer = callable.kind === reflect.MemberKind.Initializer;
+    const name = isInitializer
+      ? type.name
+      : Case.pascal(callable.name);
+
+    const params = callable.parameters.map(p => this.formatParameter(this.parameter(p))).join(', ');
+    const signatures = [`public ${name}(${params})`];
+    const invocations = [isInitializer
+      ? `new ${name}(${params});`
+      : `${type.name}.${name}(${params});`];
+
+    return {
+      name,
+      parentType: type,
+      import: this.formatImport(type),
+      parameters: callable.parameters,
```

Comment:
Would prefer to sort the parameters so that the docs look consistent between languages.

Reply:
Yeah, all parameters should be sorted by optionality:

https://github.com/cdklabs/jsii-docgen/blob/58a61628ea9b3c951b1a88304198e5e094f5f8a8/src/docgen/transpile/java.ts#L117